blog-post.md

```diff
@@ -0,0 +1,292 @@
+class `Sum` for types `a` and `b`, then we can also use function 
+`Sum.sum` on pairs of such types. The reader must noticed that
+in order to access the pair components we use **pattern matching**, 
+another feature which will be part of the Core Solidity.
+
+### Algebraic data types and pattern matching 
+
+Algebraic Data Types provide a way of modeling data modeling using sum types 
+(disjoint unions) and product types (structural records). This enables the construction of 
+precise types where invalid states are unrepresentable by design, thereby enhancing program 
+correctness through the type system itself. As an example, consider the following type 
+definition for an auction state. 
+
+```
+data AuctionState =
+    NotStarted { 
+        reservePrice: word  
+    } |
+    Active { 
+        highestBid: word,
+        leadingBidder: address 
+    } |
+    Ended { 
+        winningBid: word,
+        winner: address 
+    } |
+    Cancelled;
+```
+The type has three constructors: `NotStarted` specifies that the action have not started yet and 
+stores its reserved price, `Active` denotes that the auction has began and it stores the 
+current highest bid and the address that made such bid, `Ended` represents that the auction has 
+finished with success and it holds the highest bid and the winner address and constructor 
+`Cancelled` is used when the auction has been cancelled.
+
+Using algebraic data types, we can define function by pattern matching. As an example, 
+consider function `processAuction` which 
+Pattern matching provides structural decomposition of data types through case analysis. This 
+ensures all possible variants are handled explicitly, eliminating partial function hazards and 
+providing formal guarantees of match completeness.
+
+```
+function processAuction(state) {
+    match state {
+    |   NotStarted(reserve) => {
+            require(msg.value >= reserve, "Bid below reserve");
+            return Active(msg.value, msg.sender);
+        }
+    |   Active(currentBid, bidder) => {
+            require(msg.value > currentBid, "Bid too low");
+            transferFunds(bidder, currentBid); // refund 
+            return Active(msg.value, msg.sender);
+        }
+    |   Cancelled => {
+            revert("Auction cancelled");
+        }
+]   |   _ => {
+            return state;
+        }
+    }
+}
+```
+
+### High-order functions
+
+Functions possess first-class status within the type system, enabling their use
+as parameters, return values, and assignable entities. This facilitates the
+implementation of higher-order functions and functional composition patterns,
+enhancing expressive capability while maintaining referential transparency.
+As an example of a high-order function, let's consider `map`, which applies a 
+given function to all elements of an memory array.
+
+```
+forall T U . function map (memory(array(T)) input, transform : (T) -> U) : memory(array(U)) {
```

Comment:
I have trouble parsing this.

First `memory(array(T)) input`. It looks kinda as if you're using prefix typing here. Was it meant to be `input : memory(array(T))`?

Second `: memory(array(U))`. I think this is the return type of the function, but what makes me unsure is that it uses `:` rather than `->` as previous examples did. Is it a mistake or is it actually not a return type?